### PR TITLE
new stats to count resources

### PR DESCRIPTION
**Pull request by lohanidamodar**
## What does this PR do?

- [x] Refactor usage to use time-series data for all the metrics
- [x] Deprecate database and aggregator calculator for metrics
- [x] Deprecate maintenance by deleting the usage data (we need to save usage data longer)

@TorstenDittmann, this needs to migration to migrate all 30m stats stats into 1h stats by combining every two row.

## Test Plan

Existing usage test

## Related PRs and Issues

(If this PR is related to any other PR or resolves any issue or related to any issue, link all related PR and issues here.)

### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?

YES

### CODE REVIEW

File: app/controllers/shared/api.php [hunk 1]
```diff
@@ -404,7 +478,6 @@
         if (
             App::getEnv('_APP_USAGE_STATS', 'enabled') == 'enabled'
             && $project->getId()
-            && $mode !== APP_MODE_ADMIN // TODO: add check to make sure user is admin
```
Comment (shimonewman) [inline app/controllers/shared/api.php:407]: Removal of the !(admin) is intended?

Comment (lohanidamodar) [reply app/controllers/shared/api.php:407]: Yes indeed.

File: src/Appwrite/Usage/Calculators/TimeSeries.php [hunk 2]
```diff
@@ -14,7 +14,16 @@ class TimeSeries extends Calculator
     protected Database $database;
     protected $errorHandler;
     private array $latestTime = [];
```
Comment (eldadfux) [inline src/Appwrite/Usage/Calculators/TimeSeries.php:16]: Let's add a few docblocks in this class to better explain what we do here.

File: src/Appwrite/Usage/Stats.php [hunk 3]
```diff
@@ -80,7 +80,8 @@ public function getNamespace()
     public function submit(): void
     {
         $projectId = $this->params['projectId'] ?? '';
-        $tags = ",projectId={$projectId},version=" . App::getEnv('_APP_VERSION', 'UNKNOWN');
+        $projectInternalId = $this->params['projectInternalId'];
```
Comment (eldadfux) [inline src/Appwrite/Usage/Stats.php:83]: Let's add a few docblocks in this class to better explain what we do here.

File: app/controllers/shared/api.php [hunk 4]
```diff
@@ -158,6 +160,78 @@
         $deletes->setProject($project);
         $database->setProject($project);
 
+        $dbForProject->on(Database::EVENT_DOCUMENT_CREATE, function ($event, Document $document) use ($usage) {
+            $collection = $document->getCollection();
+            switch ($collection) {
+                case 'users':
```
[the remark this answers is not in the document]
Comment (lohanidamodar) [reply app/controllers/shared/api.php:166]: not sure I understand what you mean? We need to count different resources based on collection so I have this format.

File: app/controllers/shared/api.php [hunk 5]
```diff
@@ -158,6 +160,78 @@
...
+                                ->setParam('documents.{scope}.count.total', 1);
+                        } else {
+                            $usage->setParam('collections.{scope}.count.total', 1);
+                        }
+                    }
+                    break;
+            }
+        });
+
+        $dbForProject->on(Database::EVENT_DOCUMENT_DELETE, function ($event, Document $document) use ($usage) {
```
[the remark this answers is not in the document]
Comment (lohanidamodar) [reply app/controllers/shared/api.php:199]: based on whether it's delete or create, the value needs to be positive or negative, we might be able to extract the callback into a separate function.

File: app/tasks/maintenance.php [hunk 6]
```diff
@@ -144,11 +143,11 @@ function notifyDeleteCache($interval)
         $executionLogsRetention = (int) App::getEnv('_APP_MAINTENANCE_RETENTION_EXECUTION', '1209600');
         $auditLogRetention = (int) App::getEnv('_APP_MAINTENANCE_RETENTION_AUDIT', '1209600');
         $abuseLogsRetention = (int) App::getEnv('_APP_MAINTENANCE_RETENTION_ABUSE', '86400');
-        $usageStatsRetention30m = (int) App::getEnv('_APP_MAINTENANCE_RETENTION_USAGE_30M', '129600'); //36 hours
-        $usageStatsRetention1d = (int) App::getEnv('_APP_MAINTENANCE_RETENTION_USAGE_1D', '8640000'); // 100 days
+        $usageStatsRetention1h = (int) App::getEnv('_APP_MAINTENANCE_RETENTION_USAGE_1H', '8640000'); //100 days
```
[the remark this answers is not in the document]
Comment (lohanidamodar) [reply app/tasks/maintenance.php:146]: last we synced you said for now let's not document it, just use it. It's not even in environment variable

File: app/controllers/shared/api.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (eldadfux) [reply app/controllers/shared/api.php:199]: maybe we can pass an argument?

**Comment by lohanidamodar** [comment]
@eldadfux 
- updated doc block comments
- renamed `_APP_MAINTENANCE_RETENTION_HOURLY_USAGE` to `_APP_MAINTENANCE_RETENTION_USAGE_HOURLY`
- renamed `$hourlyUsageStatsRetention` to `$usageStatsRetentionHourly`
- renamed `setHourlyUsageRetentionDatetime` to `setUsageRetentionHourlyDateTime`

File: src/Appwrite/Event/Delete.php [hunk 7]
```diff
@@ -56,26 +55,14 @@ public function setDatetime(string $datetime): self
...
 
     /**
-     * Set datetime for 1 day interval.
+     * Sets datetime for 1h interval.
      *
      * @param string $datetime
      * @return self
      */
-    public function setDateTime1d(string $datetime): self
+    public function setDateTime1h(string $datetime): self
```
[the remark this answers is not in the document]
Comment (lohanidamodar) [reply src/Appwrite/Event/Delete.php]: updated

File: app/workers/deletes.php [hunk 8]
```diff
@@ -105,7 +105,7 @@ public function run(): void
                 break;
 
             case DELETE_TYPE_USAGE:
-                $this->deleteUsageStats($this->args['dateTime1d'], $this->args['dateTime30m']);
+                $this->deleteUsageStats($this->args['dateTime1d'], $this->args['dateTime1h']);
```
[the remark this answers is not in the document]
Comment (lohanidamodar) [reply app/workers/deletes.php]: updated

File: app/tasks/maintenance.php [hunk 9]
```diff
@@ -144,11 +143,11 @@ function notifyDeleteCache($interval)
...
+
         $cacheRetention = (int) App::getEnv('_APP_MAINTENANCE_RETENTION_CACHE', '2592000'); // 30 days
 
-        Console::loop(function () use ($interval, $executionLogsRetention, $abuseLogsRetention, $auditLogRetention, $usageStatsRetention30m, $usageStatsRetention1d, $cacheRetention) {
+        Console::loop(function () use ($interval, $executionLogsRetention, $abuseLogsRetention, $auditLogRetention, $usageStatsRetention1h, $cacheRetention) {
```
[the remark this answers is not in the document]
Comment (lohanidamodar) [reply app/tasks/maintenance.php]: updated

File: app/config/variables.php [hunk 10]
```diff
@@ -188,7 +188,7 @@
             ],
             [
                 'name' => '_APP_USAGE_DATABASE_INTERVAL',
-                'description' => 'Interval value containing the number of seconds that the Appwrite usage process should wait before aggregating stats from data in Appwrite Database. The default value is 15 minutes.',
+                'description' => 'Deprecated since _ use _APP_USAGE_AGGREGATION_INTERVAL instead.',
```
[the remark this answers is not in the document]
Comment (lohanidamodar) [reply app/config/variables.php:191]: set

File: app/config/variables.php [hunk 11]
```diff
@@ -179,7 +179,7 @@
             ],
             [
                 'name' => '_APP_USAGE_TIMESERIES_INTERVAL',
-                'description' => 'Interval value containing the number of seconds that the Appwrite usage process should wait before aggregating stats and syncing it to Appwrite Database from Timeseries Database. The default value is 30 seconds.',
+                'description' => 'Deprecated since _ use _APP_USAGE_AGGREGATION_INTERVAL instead.',
```
[the remark this answers is not in the document]
Comment (lohanidamodar) [reply app/config/variables.php:182]: set

File: app/controllers/shared/api.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (lohanidamodar) [reply app/controllers/shared/api.php:199]: refactored into single function

File: composer.json [hunk 12]
```diff
@@ -51,7 +51,7 @@
         "utopia-php/cache": "0.6.*",
         "utopia-php/cli": "0.13.*",
         "utopia-php/config": "0.2.*",
-        "utopia-php/database": "0.26.*",
+        "utopia-php/database": "0.27.0 as 0.26.0",
```
[the remark this answers is not in the document]
Comment (lohanidamodar) [reply composer.json:54]: Working on other depending libraries upgrading (abuse, audit)

File: tests/e2e/General/HTTPTest.php [hunk 13]
```diff
@@ -163,7 +163,8 @@ public function testSpecs()
 
             $response['body'] = json_decode($response['body'], true);
             $this->assertEquals(200, $response['headers']['status-code']);
-            $this->assertEmpty($response['body']['schemaValidationMessages']);
+            // looks like recent change in the validator
```
[the remark this answers is not in the document]
Comment (lohanidamodar) [reply tests/e2e/General/HTTPTest.php:166]: synced and updated

File: app/tasks/usage.php [hunk 14]
```diff
@@ -114,63 +110,28 @@ function getInfluxDB(Registry &$register): InfluxDatabase
...
-        Console::info("[{$now}] Aggregation took {$loopTook} seconds");
-    }, $interval);
-}
-
 $cli
     ->task('usage')
-    ->param('type', 'timeseries', new WhiteList(['timeseries', 'database']))
     ->desc('Schedules syncing data from influxdb to Appwrite console db')
-    ->action(function (string $type) use ($register, $logError) {
+    ->action(function () use ($register, $logError) {
```
[the remark this answers is not in the document]
Comment (lohanidamodar) [reply app/tasks/usage.php:116]: updated with DI

File: composer.json (hunk 12 above)
[the remark this answers is not in the document]
Comment (lohanidamodar) [reply composer.json]: Updated with latest versions

**Review by eldadfux — APPROVED** [review]

File: app/config/collections.php [hunk 15]
```diff
@@ -2327,6 +2327,17 @@
...
+                '$id' => ID::custom('buildTime'),
+                'type' => Database::VAR_INTEGER,
+                'format' => '',
+                'size' => 0,
+                'signed' => true,
+                'required' => false,
+                'default' => null,
+                'array' => false,
+                'filters' => [],
+            ],
```
Comment (TorstenDittmann) [inline app/config/collections.php:2340]: Why did we add this? This should be part of the build itself 👍🏻

File: app/controllers/shared/api.php [hunk 16]
```diff
@@ -47,6 +48,47 @@
...
+        default:
+            if (strpos($collection, 'bucket_') === 0) {
+                $usage
+                    ->setParam('bucketId', $document->getAttribute('bucketId'))
+                    ->setParam('files.{scope}.storage.size', $document->getAttribute('sizeOriginal') * $multiplier)
+                    ->setParam('files.{scope}.count.total', 1 * $multiplier);
+            } elseif (strpos($collection, 'database_') === 0) {
+                $usage
+                    ->setParam('databaseId', $document->getAttribute('databaseId'));
+                if (strpos($collection, '_collection_') != false) {
```
Comment (TorstenDittmann) [inline app/controllers/shared/api.php:80]:
```suggestion
                if (strpos($collection, '_collection_') !== false) {
```

File: src/Appwrite/Usage/Stats.php [hunk 17]
```diff
@@ -102,12 +105,13 @@ public function submit(): void
...
             'users.{scope}.requests.update',
             'users.{scope}.requests.delete',
+            'users.{scope}.count.total',
         ];
 
         foreach ($usersMetrics as $metric) {
             $value = $this->params[$metric] ?? 0;
-            if ($value >= 1) {
-                $this->statsd->increment($metric . $tags);
+            if ($value == 1 || $value == -1) {
```
Comment (TorstenDittmann) [inline src/Appwrite/Usage/Stats.php:113]:
```suggestion
            if ($value === 1 || $value === -1) {
```

File: src/Appwrite/Usage/Stats.php [hunk 18]
```diff
@@ -124,13 +128,16 @@ public function submit(): void
...
             'documents.{scope}.requests.delete',
+            'databases.{scope}.count.total',
+            'collections.{scope}.count.total',
+            'documents.{scope}.count.total'
         ];
 
         foreach ($dbMetrics as $metric) {
             $value = $this->params[$metric] ?? 0;
-            if ($value >= 1) {
+            if ($value == 1 || $value == -1) {
```
Comment (TorstenDittmann) [inline src/Appwrite/Usage/Stats.php:138]:
```suggestion
            if ($value === 1 || $value === -1) {
```

File: src/Appwrite/Usage/Stats.php [hunk 19]
```diff
@@ -143,13 +150,16 @@ public function submit(): void
...
             'files.{scope}.requests.delete',
+            'buckets.{scope}.count.total',
+            'files.{scope}.count.total',
+            'files.{scope}.storage.size'
         ];
 
         foreach ($storageMertics as $metric) {
             $value = $this->params[$metric] ?? 0;
-            if ($value >= 1) {
+            if ($value >= 1 || $value == -1 || $value < -1) {
```
Comment (TorstenDittmann) [inline src/Appwrite/Usage/Stats.php:160]:
```suggestion
            if ($value !== 0) {
```
?

File: src/Appwrite/Usage/Stats.php [hunk 20]
```diff
@@ -176,19 +186,30 @@ public function submit(): void
...
         $functionCompute = $functionExecutionTime + $functionBuildTime;
+        $functionTags = $tags . ',functionId=' . $functionId;
+
+        $deploymentSize = $this->params['deployment.{scope}.storage.size'] ?? 0;
+        $storageSize = $this->params['files.{scope}.storage.size'] ?? 0;
+        if ($deploymentSize + $storageSize > 0 || $deploymentSize + $storageSize <= -1) {
+            $this->statsd->count('project.{scope}.storage.size' . $tags, $deploymentSize + $storageSize);
+        }
+
+        if ($deploymentSize > 0 || $deploymentSize <= -1) {
```
Comment (TorstenDittmann) [inline src/Appwrite/Usage/Stats.php:197]:
```suggestion
        if ($deploymentSize !== 0) {
```

is the same right?

File: app/config/collections.php (hunk 15 above)
Comment (lohanidamodar) [reply app/config/collections.php]: fixed

**Review by TorstenDittmann — APPROVED** [review]

**Pull request merged**